lib/spack/spack/installer.py

```
@@ -1428,6 +1432,9 @@ def install(self, **kwargs):
             # assume using a separate (failed) prefix lock file.
             if pkg_id in self.failed or spack.store.db.prefix_failed(spec):
                 tty.warn('{0} failed to install'.format(pkg_id))
+                if fail_fast:
+                    raise InstallError('Fail-fast termination of install')
+
                 self._update_failed(task)
```

(question) Should the `InstallError` be raised after this to alert other processes that there was a failure?

---

Excellent point.  I debated where to place raising the error.

I rationalized that a user who wants to fail fast wants to terminate on the first package.  If there is a concurrent, overlapping installation then the worst the other install would encounter would be to try to install the package again.  If there isn't a concurrent, overlapping installation then all of the failure marks are wiped out on the next install, which has to detect the failure on its own anyway.

---

> If there is a concurrent, overlapping installation then the worst the other install would encounter would be to try to install the package again

I agree, although I think that can also be prevented generally by adding the failure prefix. Also I'm thinking that separate processes may keep trying to install packages as long as they don't try the one that failed. Consider e.g.

```
 W
/ \
X  Y
   |
   Z
```

If there are two processes, P1, and P2, and P1 fails on X while P2 succeeds with Z, would it continue to try installing Y without the failure prefix?

---

Changed